pkg/node/manager/node_config_notifier.go

```
@@ -0,0 +1,43 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package manager
+
+import (
+	"errors"
+
+	datapath "github.com/cilium/cilium/pkg/datapath/types"
+)
+
+type NodeConfigNotifier struct {
+	handlers []datapath.NodeConfigChangeHandler
+}
+
+func newNodeConfigNotifier() *NodeConfigNotifier {
+	return &NodeConfigNotifier{
+		handlers: make([]datapath.NodeConfigChangeHandler, 0),
+	}
+}
+
+func (n *NodeConfigNotifier) Subscribe(handler datapath.NodeConfigChangeHandler) {
+	n.handlers = append(n.handlers, handler)
```

How is it guaranteed that this function is not called from multiple go routines concurrently?

If the reason is that it is intended to only be used used in Hive constructors (which are guaranteed to be sequential), why not use a Hive `group`? In any case, if this is kept either document that this is not thread-safe or add a mutex.

---

That is exactly the reason. Its only meant to be called from hive constructors.

>  why not use a Hive `group`?

I think consensus amongst sig-foundation is that having a observer pattern like this is desirable over using a `group`  unless you have a good reason to use groups.

Explicitly subscribing via a method makes it easier to find the dependency. With groups you have to search for functions returning the handler interface, a more difficult query. The registrar (`NodeConfigNotifier` in this case) is also explicitly added to the parameters of call constructors that so something with it. 

A group is still a good option when manual registration would add a disproportional burden. Take metrics for example, having them be collected in a single group automatically instead of manually registering tens of not hundreds of individual metrics makes it worth it.

Another good reason for groups over explicit registration is the dependency order. In this example, the `NodeConfigNotifier` has no dependencies, and both the handlers and caller depend on `NodeConfigNotifier`.

With a group, the `NodeConfigNotifier` would depend on the handlers, and the caller on the `NodeConfigNotifier`, which makes it more likely that you accidentally create a dependency cycle, which is impossible with explicit registration when done in this way.

\</rambling>

> In any case, if this is kept either document that this is not thread-safe or add a mutex.

Yes, very good idea. I will also detection for modification after construction time, since that would also be unexpected.

---

Neat, thanks for changing it! 

>  I will also detection for modification after construction time, since that would also be unexpected.

This is celver. I do think however have used a log line (rather than a error) would be much better. The error will be ignored by callers (in fact, the current callers do also ignore it) and thus it might never be seen by a developer. Having a log line with level warning will also make sure CI fails if someone adds this.

---

Yes, absolutely right.